include/bout/petsc_interface.hxx

```diff
@@ -0,0 +1,600 @@
+  PetscInt getGlobal(Ind3D ind);
+  PetscInt getGlobal(IndPerp ind);
+
+protected:
+  GlobalIndexer(Mesh* localmesh);
+  Mesh* fieldmesh;
+
+private:
+  /// This gets called by initialiseTest and is used to register
+  /// fields with fake parallel meshes.
+  virtual void registerFieldForTest(FieldData& f);
+  virtual void registerFieldForTest(FieldPerp& f);
+
+  PetscLib lib;
+
+  /// Fields containing the indices for each element (as reals)
+  Field3D indices3D;
+  Field2D indices2D;
+  FieldPerp indicesPerp;
+
+  /// The only instance of this class acting on the global Mesh
+  static IndexerPtr globalInstance;
+  static bool initialisedGlobal;
+  static Mesh* globalmesh;
+  bool initialised;
+};
+
+/*!
+ * A class which wraps PETSc vector objects, allowing them to be
+ * indexed using the BOUT++ scheme. Note that boundaries are only
+ * included in the vector to a depth of 1.
+ */
+template <class T>
+class PetscVector;
+template <class T>
+void swap(PetscVector<T>& first, PetscVector<T>& second);
+
+template <class T>
+class PetscVector {
+public:
+  static_assert(bout::utils::is_Field<T>::value, "PetscVector only works with Fields");
+  using ind_type = typename T::ind_type;
+
+  struct VectorDeleter {
+    void operator()(Vec* v) const {
+      VecDestroy(v);
+      delete v;
+    }
+  };
+
+  /// Default constructor does nothing
+  PetscVector() : vector(new Vec(), VectorDeleter()) {}
+
+  /// Copy constructor
+  PetscVector(const PetscVector<T>& v) : vector(new Vec(), VectorDeleter()) {
+    VecDuplicate(*v.vector, vector.get());
+    VecCopy(*v.vector, *vector);
+    indexConverter = v.indexConverter;
+    location = v.location;
+    initialised = v.initialised;
+  }
+
+  /// Move constrcutor
+  PetscVector(PetscVector<T>&& v) {
+    std::swap(vector, v.vector);
+    indexConverter = v.indexConverter;
+    location = v.location;
+    initialised = v.initialised;
+    v.initialised = false;
+  }
+
+  /// Construct from a field, copying over the field values
+  PetscVector(const T& f) : vector(new Vec(), VectorDeleter()) {
+    const MPI_Comm comm =
+        std::is_same<T, FieldPerp>::value ? f.getMesh()->getXcomm() : BoutComm::get();
+    indexConverter = GlobalIndexer::getInstance(f.getMesh());
+    const auto size = [&f]() {
+      if (std::is_same<T, FieldPerp>::value) {
+        return f.getMesh()->localSizePerp();
+      } else if (std::is_same<T, Field2D>::value) {
+        return f.getMesh()->localSize2D();
+      } else if (std::is_same<T, Field3D>::value) {
+        return f.getMesh()->localSize3D();
+      } else {
+        throw BoutException("PetscVector initialised for non-field type.");
+      }
+    };
+    VecCreateMPI(comm, size(), PETSC_DECIDE, vector.get());
+    location = f.getLocation();
+    initialised = true;
+    BOUT_FOR(i, f.getRegion(RGN_ALL)) {
+      PetscInt ind = indexConverter->getGlobal(i);
+      if (ind != -1) {
+        BOUT_OMP(critical) VecSetValues(*vector, 1, &ind, &f[i], INSERT_VALUES);
```

Comment:
Either would work. I was thinking that since there are a few non-critical statements in the loop parallelism might give some performance boost but on reflection probably not (given the overhead involved).

I wasn't aware of `VecSetValue` as it wasn't mentioned in the PDF PETSc manual. I've switched over to that where appropriate. Longer term, it may be possible to optimise this bit of code by using `VecSetValues` wherever there is a continuous stretch of field data which should be copied.